Suicide Revamp

Domic2 opened the pull request:
:cl: Domic
tweak: Suicide now accounts for damage, what you're wearing, where your brain actually is, utilizes aim mode, and is now triggered manually instead of being on a timer.
bugfix: People who suicide are no longer save-able via MMI.
bugfix: Suicide via less-than-lethal weapons is no longer easy to to.
/:cl:

In addition to what's stated in the changelog, I've added a proc which checks if a body part is covered with something, stopped the brain from giving brain damage messages after death, permitted suicide to use targeting triggers, and prevented suicide victims from being recoverable via MMI.

I'd just like to clarify that suicide is still (HEAVILY) discouraged on the server, the purpose of this PR is to make it less underwhelming and more realistic (no instant disorientator suicide, please) when used properly.

A side effect to making brain damage 'easier' to deal is psionic awakening potentially being easier to do.

Domic2 committed:
Suicide improvements

SierraKomodo commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -116,6 +116,37 @@ meteor_act
 			return gear
 	return null
 
+//Takes a body part and checks if it's covered
```
https://github.com/Baystation12/Baystation12/wiki/Guide-to-SpacemanDMM-Docblocks

```suggestion
/// Takes a body part and checks if it's covered
```

Domic2 commented:
Last force push fixed the following:
Human_defense.dm line 126 was: 'if ("face")'. Checking a BP never actually gives "face". Changed to: 'if ("mouth")'.
Gun.dm line 482 changed from 'bodypart = "face"' to 'bodypart = "mouth"' to reflect this.

Domic2 commented:
human_defense.dm: Used the || operator instead of a comma in a switch statement :(
Now fixed.

PsiOmegaDelta commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -116,6 +116,37 @@ meteor_act
+//Takes a body part and checks if it's covered
+/mob/living/carbon/human/proc/check_bp_coverage(var/bodypart)
+	switch(bodypart)//Let's convert it to something we can use first
+		if ("head")
```
BP_HEAD

PsiOmegaDelta commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -116,6 +116,37 @@ meteor_act
+	switch(bodypart)//Let's convert it to something we can use first
+		if ("head")
+			bodypart = HEAD
+		if ("eyes")
```
BP_EYES

PsiOmegaDelta commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -116,6 +116,37 @@ meteor_act
+		if ("groin")
+			bodypart = LOWER_TORSO
+		if ("l_arm", "r_arm")
+			bodypart =  ARMS
```
Etc.
You should generally never use plain, non-constant/non-define strings in comparisons

PsiOmegaDelta commented:
Will note that I don't feel that suicide code should even be a thing.
Indeed it should be covered in ageneral scenario of trying to harm a help/defenseless mob.

Domic2 commented:
I used those before and they didn't work, re-testing now shows different results, I guess I did something wrong back then.

I'm not sure how I feel about the general scenario idea, I think I prefer this one. Probably a bit biased on this because I spent an embarrassing amount of time trying to shoot myself for this code to work, though

SierraKomodo commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -116,6 +116,37 @@ meteor_act
+	var/list/protective_gear = list(head, wear_mask, wear_suit, w_uniform, gloves, shoes, glasses)
+	for(var/bp in protective_gear)
+		if(!bp)	continue
+		if(bp && istype(bp ,/obj/item/clothing))
```
These two if statements are entirely redundant. Istype also handles nullchecking (`null` is not a type of `/obj/item/clothing`). I'd suggest one of these two options:

```dm
	if (!istype(bp, /obj/item/clothing))
		continue
	// Rest of your loop code here
```

or

```dm
	if (istype(bp, /obj/item/clothing))
		// Rest of your loop code here
```

SierraKomodo commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -116,6 +116,37 @@ meteor_act
 	return null
 
+//Takes a body part and checks if it's covered
+/mob/living/carbon/human/proc/check_bp_coverage(var/bodypart)
```
```suggestion
/mob/living/carbon/human/proc/check_bp_coverage(bodypart)
```

SierraKomodo commented on code/modules/organs/internal/brain.dm:
```diff
@@ -129,11 +129,11 @@
 	to_chat(owner, "<span class = 'notice' font size='10'><B>Where am I...?</B></span>")
 	sleep(5 SECONDS)
-	if(!owner)
+	if(!owner || damage >= max_damage)
```
I feel that death state checks would be a better option than a flat damage check. The below suggestion checks both if the owner is considered dead and if the brain organ itself is considered dead.

```suggestion
	if(!owner || owner.stat == DEAD || (status & ORGAN_DEAD))
```

SierraKomodo commented on code/modules/organs/internal/brain.dm:
```diff
@@ -129,11 +129,11 @@
 	to_chat(owner, "<span class = 'notice' font size='10'><B>What's going on...?</B></span>")
 	sleep(10 SECONDS)
-	if(!owner)
+	if(!owner || damage >= max_damage)
```
```suggestion
	if(!owner || owner.stat == DEAD || (status & ORGAN_DEAD))
```

SierraKomodo commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
-			return
+	if (world.time < next_fire_time)
+		if (world.time % 3) //to prevent spam
+			to_chat(M, SPAN_WARNING("[src] is not ready to fire again!"))
```
```suggestion
			to_chat(M, SPAN_WARNING("\The [src] is not ready to fire again!"))
```

SierraKomodo commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
+
+	last_safety_check = world.time
+	admin_attacker_log(M, "is trying to commit suicide with \a [src]")
+	user.visible_message(M, SPAN_WARNING("[M] pulls the trigger."))
```
```suggestion
	user.visible_message(M, SPAN_WARNING("\The [M] pulls the trigger."))
```

SierraKomodo commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
+		if (!in_chamber)
+			handle_click_empty(M)
+			break
+		to_chat(M, SPAN_NOTICE("You feel \the [src] go off..."))
```
Unsure this is the best place for this message - It'll result in spam that sends a message every shot in a burst.

Domic2 replied:
That was intended, but it's probably for the best to move it before the for statement.

SierraKomodo commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
+				else
+					dmgmultiplier = 0.5
+				if (blocked)
+					to_chat(M, SPAN_WARNING("A clear shot to your [bodypart] is blocked by the [blocked], significantly reducing damage to the [brain.name]!"))
```
```suggestion
					to_chat(M, SPAN_WARNING("A clear shot to your [bodypart] is blocked by \the [blocked], significantly reducing damage to \the [brain.name]!"))
```

SierraKomodo commented on code/modules/projectiles/targeting/targeting_overlay.dm:
```diff
@@ -24,6 +24,12 @@
 	loc = null
 	verbs.Cut()
 
+/obj/aiming_overlay/proc/toggle_permission_silent(var/perm)
+	if (target_permissions & perm)
+		target_permissions &= ~perm
+	else
+		target_permissions |= perm
+
```
Some copy paste here. In `toggle_permission()` I'd suggest replacing the original `target_permissions` code with `toggle_permission_silent(perm)` - Alternatively, add a `silent` parameter to `toggle_permission` and, if `silent == TRUE`, halt the proc after updating `target_permissions`.

```dm
/obj/aiming_overlay/proc/toggle_permission(perm, silent)

	if(target_permissions & perm)
		target_permissions &= ~perm
	else
		target_permissions |= perm

	if (silent)
		return
```

SierraKomodo commented on code/modules/projectiles/targeting/targeting_overlay.dm:
```diff
@@ -149,17 +156,29 @@ obj/aiming_overlay/proc/update_aiming_deferred()
 		to_chat(owner, "<span class='warning'>You cannot aim a gun while handcuffed.</span>")
 		return
 
+	var/gunpointedtarget = "<span class='danger'>You now have a gun pointed at you. No sudden moves!</span>"
```
```suggestion
	var/gunpointedtarget = SPAN_DANGER("You now have a gun pointed at you. No sudden moves!")
```

SierraKomodo replied:
```suggestion
	var/gunpointedtarget = "<span class='danger'>You now have \a [thing] pointed at you. No sudden moves!</span>"
```

SierraKomodo commented on code/modules/projectiles/targeting/targeting_overlay.dm:
```diff
@@ -149,17 +156,29 @@ obj/aiming_overlay/proc/update_aiming_deferred()
+	var/gunpointedself = ""
+	var/mob/living/carbon/human/user = owner
+	if (istype(user) && user.check_has_mouth() && !(user.check_mouth_coverage()))
+		gunpointedself = "<span class='danger'>\The [owner] puts the barrel of \the [thing] in their mouth, ready to pull the trigger...</span>"
```
```suggestion
		gunpointedself = SPAN_DANGER("\The [owner] puts the barrel of \the [thing] in their mouth, ready to pull the trigger...")
```

SierraKomodo commented on code/modules/projectiles/targeting/targeting_overlay.dm:
```diff
@@ -149,17 +156,29 @@ obj/aiming_overlay/proc/update_aiming_deferred()
+	if (istype(user) && user.check_has_mouth() && !(user.check_mouth_coverage()))
+		gunpointedself = "<span class='danger'>\The [owner] puts the barrel of \the [thing] in their mouth, ready to pull the trigger...</span>"
+	else
+		gunpointedself = "<span class='danger'>\The [owner] aims \the [thing] at themselves, ready to pull the trigger...</span>"
```
```suggestion
		gunpointedself = SPAN_DANGER(">\The [owner] aims \the [thing] at themselves, ready to pull the trigger...")
```

Domic2 commented:
That's what I get for carelessly copy-pasting stuff...
Additional changes include:
Also removed redundant checks from human_defense.dm line 104.
I previously added die() to line 220 of Brain.dm. I do not know why I did this, so it's gone now.
Changed to_chat stuff in targeting_overlay.dm line 75, found out the messages were basically saying the same thing twice during testing.
Added spaces to random if statements.

SierraKomodo reviewed (commented):
Not convinced this should auto-switch to aim-mode when you target mouth and click yourself. It should instead just shoot if you're not on aim mode, like it does when targeting the head.

SierraKomodo reviewed (changes requested):
This also appears to have broken the currently functionality of aiming at yourself when targeting any body part except mouth.

Domic2 commented:
Regarding the auto-switch, if someone doesn't look at the changelog and tries to suicide, or threaten a suicide, via the old method, it would produce less-than-desirable results. Head damage can be achieved much more easily by selecting the head instead, too.
If you don't mind me asking, what purpose would allowing aim mode on self serve? It seemed somewhat pointless (and not well thought out considering all the double messages) so I set it to suicide only

Spookerton commented:
It allows you to aim at yourself. For dramatic effect, y'know.

Spookerton commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -101,8 +101,7 @@ meteor_act
 
 	var/list/body_parts = list(head, wear_mask, wear_suit, w_uniform)
 	for(var/bp in body_parts)
-		if(!bp)	continue
-		if(bp && istype(bp ,/obj/item/clothing))
+		if(istype(bp ,/obj/item/clothing))
 			var/obj/item/clothing/C = bp
 			if(C.body_parts_covered & HEAD)
 				return 1
```
Same as other loop, housekeeping since we're touching it anyway
```
for (var/obj/item/clothing/C in list(head, wear_mask, wear_suit, w_uniform))
	if (C.body_parts_covered & HEAD)
		return TRUE
```

Spookerton commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -116,6 +115,36 @@ meteor_act
+		if (BP_L_FOOT, BP_R_FOOT)
+			bodypart = FEET
+
+	var/list/protective_gear = list(head, wear_mask, wear_suit, w_uniform, gloves, shoes, glasses)
+	for(var/bp in protective_gear)
+		if(istype(bp ,/obj/item/clothing))
+			var/obj/item/clothing/C = bp
+			if(C.body_parts_covered & bodypart)
+				return bp
+	return null
```
```suggestion
	for (var/obj/item/clothing/C in list(head, wear_mask, wear_suit, w_uniform, gloves, shoes, glasses))
		if (C.body_parts_covered & bodypart)
			return C
```
inherent istype in typed loop, no need to explicitly return null

Spookerton commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -116,6 +115,36 @@ meteor_act
 
+///Takes a body part and checks if it's covered
+/mob/living/carbon/human/proc/check_bp_coverage(bodypart)
+	switch(bodypart)//Let's convert it to something we can use first
```
```suggestion
	switch(bodypart)
```

Spookerton commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -116,6 +115,36 @@ meteor_act
 			return gear
 	return null
 
+///Takes a body part and checks if it's covered
+/mob/living/carbon/human/proc/check_bp_coverage(bodypart)
```
```suggestion
/// Returns null or the first equipped item covering BP_* bodypart.
/mob/living/carbon/human/proc/get_clothing_covering(bodypart)
```
not a check - used to fetch the covering thing

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
+	var/bodypart = brain.parent_organ
+	if (brain.parent_organ == BP_HEAD)
+		bodypart = BP_MOUTH
+	var/obj/item/blocked = M.check_bp_coverage(bodypart)
```
```suggestion
	var/obj/item/blocked = M.get_clothing_covering(bodypart)
```

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -184,14 +184,33 @@
+			admin_attacker_log(M, "is getting ready to suicide with \a [src]")
+			if (M.aiming.aiming_at != M)
+				var/checkperm = FALSE
+				if (!(M.aiming.target_permissions & TARGET_CAN_CLICK)) // let's not kill ourselves immidiately
```
```suggestion
				if (!(M.aiming.target_permissions & TARGET_CAN_CLICK))
```

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
+		if (world.time % 3) //to prevent spam
+			to_chat(M, SPAN_WARNING("\The [src] is not ready to fire again!"))
+		return
+	M.setClickCooldown((burst - 1)* burst_delay) //no clicking on things while shooting yourself, moving is fine though
```
```suggestion
	M.setClickCooldown((burst - 1) * burst_delay)
```

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
-			mouthshoot = 0
-			return
+	if (world.time < next_fire_time)
+		if (world.time % 3) //to prevent spam
```
```suggestion
		if (world.time % 3)
```

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -184,14 +184,33 @@
+				handle_suicide(user)
+			return
+		else if (user.aiming && user.aiming.active)
+			return //aim mode on self is suicide only.
```
```suggestion
			return
```

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
+		bodypart = BP_MOUTH
+	var/obj/item/blocked = M.check_bp_coverage(bodypart)
+
+	for(var/i in 1 to burst)
```
```suggestion
	for (var/i = 1 to burst)
```

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
-			user.death()
+			in_chamber.on_hit(M, 0, brain.parent_organ)
+			if (istype(in_chamber, /obj/item/projectile/ion))
+				in_chamber.on_impact(M) //If applied to everything it tends to leave projectiles floating in the air
```
```suggestion
				in_chamber.on_impact(M)
```

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
+					dmgmultiplier = 0.5
+				if (blocked)
+					to_chat(M, SPAN_WARNING("A clear shot to your [bodypart] is blocked by the [blocked], significantly reducing damage to \the [brain.name]!"))
+					dmgmultiplier = dmgmultiplier/5 //If you're trying to kill yourself, it's fair to assume you will try to aim the gun optimally, so armor won't have a significant impact, but the damage will still be dampened.
```
```suggestion
					dmgmultiplier = dmgmultiplier / 5
```

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
+				M.apply_damage(in_chamber.damage*2, in_chamber.damage_type, brain.parent_organ, in_chamber.damage_flags(), used_weapon = "Point blank shot in the mouth with \a [in_chamber]")
+				var/dmgmultiplier
+				if (prob (95))
+					dmgmultiplier = rand(30, 50)/10
```
```suggestion
					dmgmultiplier = rand(30, 50) / 10
```

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
+			if (istype(in_chamber, /obj/item/projectile/ion))
+				in_chamber.on_impact(M) //If applied to everything it tends to leave projectiles floating in the air
+			if (in_chamber.damage != 0)
+				M.apply_damage(in_chamber.damage*2, in_chamber.damage_type, brain.parent_organ, in_chamber.damage_flags(), used_weapon = "Point blank shot in the mouth with \a [in_chamber]")
```
```suggestion
				M.apply_damage(in_chamber.damage * 2, in_chamber.damage_type, brain.parent_organ, in_chamber.damage_flags(), used_weapon = "Point blank shot in the mouth with \a [in_chamber]")
```

Spookerton commented on code/modules/projectiles/gun.dm:
```diff
@@ -439,47 +458,69 @@
+			to_chat(M, SPAN_WARNING("\The [src] is not ready to fire again!"))
+		return
+	M.setClickCooldown((burst - 1)* burst_delay) //no clicking on things while shooting yourself, moving is fine though
+	next_fire_time = world.time + max(burst_delay+1, fire_delay)
```
```suggestion
	next_fire_time = world.time + max(burst_delay + 1, fire_delay)
```

Spookerton commented on code/modules/projectiles/targeting/targeting_overlay.dm:
```diff
@@ -136,33 +141,45 @@ obj/aiming_overlay/proc/update_aiming_deferred()
+		if (owner != target)
+			owner.visible_message(SPAN_DANGER("\The [owner] turns \the [thing] on \the [target]!"))
+			to_chat(target, "[gunpointedtarget]")
+		else owner.visible_message("[gunpointedself]")
```
```suggestion
		else
			owner.visible_message("[gunpointedself]")
```

Spookerton commented on code/modules/projectiles/targeting/targeting_overlay.dm:
```diff
@@ -136,33 +141,45 @@ obj/aiming_overlay/proc/update_aiming_deferred()
+	else if (owner != target)
+		owner.visible_message(SPAN_DANGER("\The [owner] aims \the [thing] at \the [target]!"))
+		to_chat(target, "[gunpointedtarget]")
+	else owner.visible_message("[gunpointedself]")
```
```suggestion
	else
		owner.visible_message("[gunpointedself]")
```

Spookerton commented on code/modules/projectiles/targeting/targeting_triggers.dm:
```diff
@@ -13,19 +13,23 @@
+	if (!istype(G))
+		return
+	if (owner == aiming_at)
+		addtimer(CALLBACK(G, /obj/item/gun/proc/handle_suicide, owner, 2)) //give enough time for them to actually do or start doing what they want their last action to be
```
```suggestion
		addtimer(CALLBACK(G, /obj/item/gun/proc/handle_suicide, owner, 2))
```

SierraKomodo reviewed (changes requested):
You marked a bunch of reviews as resolved but didn't actually push an update to the PR.

Also, you have merge conflicts now.

Domic2 commented:
Ah, sorry, everything except re-enabling aim mode on self is done, so I decided not to push it till it's handled

Domic2 commented:
Okay, merge conflict resolved.
All requested changes are done, and you can now aim at yourself with no problem.
Changing the targeted bodypart from/to the mouth when aiming at yourself gives feedback to account for this

Domic2 commented:
Merge conflict with #30811 resolved. Just my luck, I guess
